### PR TITLE
Moved calculation of mse to common_step and added logging of mean rmse

### PR DESCRIPTION
**Pull request by mfroelund**
## Describe your changes

Logging rmse during training, validation and testing.

## Issue Link

< Link to the relevant issue or task. > (e.g. `closes #00` or `solves #00`)

## Type of change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [ ] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [ ] I have performed a self-review of my code
- [ ] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [ ] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [ ] the code is readable
- [ ] the code is well tested
- [ ] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain

## Author checklist after completed review

- [ ] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug

## Checklist for assignee

- [ ] PR is up to date with the base branch
- [ ] the tests pass
- [ ] author has added an entry to the changelog (and designated the change as *added*, *changed* or *fixed*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

File: neural_lam/models/ar_model.py [hunk 1]
```diff
@@ -283,13 +283,31 @@ def common_step(self, batch):
...
+        # Calculate MSEs
+        entry_mses = metrics.mse(
+            prediction,
+            target_states,
+            pred_std,
+            mask=self.interior_mask_bool,
+            sum_vars=False,
+        )  # (B, pred_steps, d_f)
+
+        # Log mean RMSE for first prediction step
```
Comment (matschreiner) [inline neural_lam/models/ar_model.py]: Redundant comment?

File: neural_lam/models/ar_model.py [hunk 2]
```diff
@@ -283,13 +283,31 @@ def common_step(self, batch):
         # prediction: (B, pred_steps, num_grid_nodes, d_f) pred_std: (B,
         # pred_steps, num_grid_nodes, d_f) or (d_f,)
 
-        return prediction, target_states, pred_std, batch_times
+        # Calculate MSEs
```
Comment (matschreiner) [inline neural_lam/models/ar_model.py]: Redundant comment?

File: neural_lam/models/ar_model.py [hunk 3]
```diff
@@ -405,10 +416,12 @@ def test_step(self, batch, batch_idx):
             batch_size=batch[0].shape[0],
         )
 
+        # Store already computed MSEs
```
Comment (matschreiner) [inline neural_lam/models/ar_model.py]: :)

**Comment by matschreiner** [comment]
LGTM. 
Just added comments about comments.
Maybe its a question about personal style, but still, here's a nice recap of the chapter about comments from the famous book "Clean Code" by uncle bob :) 
https://medium.com/codex/clean-code-comments-833e11a706dc

**Comment by mfroelund** [comment]
Nice, thanks for the link. Have you added comments to the comments in this PR? I don't see the comments then:(

**Comment by matschreiner** [comment]
@mafdmi  Could it be possible to also log the learning rate during training? 
I think you can find it on 

`model.trainer.optimizer.param_groups[0]["lr"]`

**Comment by matschreiner** [comment]
> Nice, thanks for the link. Have you added comments to the comments in this PR? I don't see the comments then:(

i did, but I don't know why you don't see them. Anyways, I just wrote that the comments were maybe redundant. :)

**Comment by mfroelund** [comment]
I've added logging of learning rate, so I think this is ready for final review. You can see the results of my test run at https://localhost:4433/#/experiments/7/runs/6f7eacff595a4515b1892029c368052b/model-metrics

**Comment by matschreiner** [comment]
LGTM! :)

**Comment by mfroelund** [comment]
@matschreiner I've fixed the tests, but for some reason two of the workflow jobs haven't been run - they have been waiting for a runner to pick them up since yesterday. Tried to re-run them today: https://github.com/mafdmi/neural-lam/actions/runs/13700752255

File: neural_lam/models/ar_model.py [hunk 4]
```diff
@@ -269,7 +269,7 @@ def unroll_prediction(self, init_states, forcing_features, true_states):
 
     def common_step(self, batch):
         """
-        Predict on single batch batch consists of: init_states: (B, 2,
+        Predict on single batch consists of: init_states: (B, 2,
```
Comment (matschreiner) [inline neural_lam/models/ar_model.py]: I don't think this docstring is clear.

File: neural_lam/models/ar_model.py [hunk 5]
```diff
@@ -283,13 +283,24 @@ def common_step(self, batch):
...
+        )  # (B, pred_steps, d_f)
+
+        return prediction, target_states, pred_std, batch_times, entry_mses
 
     def training_step(self, batch):
         """
         Train on single batch
         """
-        prediction, target, pred_std, _ = self.common_step(batch)
+        prediction, target, pred_std, _, entry_mses = self.common_step(batch)
```
Comment (matschreiner) [inline neural_lam/models/ar_model.py]: I think the common step should have a more descriptive name - it should reflect it's function rather than the fact that it is being shared. 
Also it has two responsibilities which is prediction with the model and processsing of the prediction - could maybe be factored into separate steps?

File: neural_lam/models/ar_model.py [hunk 6]
```diff
@@ -298,9 +309,18 @@ def training_step(self, batch):
             )
         )  # mean over unrolled times and batch
 
-        log_dict = {"train_loss": batch_loss}
+        # Logging
+        train_log_dict = {"train_loss": batch_loss}
+        state_var_names = self._datastore.get_vars_names(category="state")
+        train_log_dict |= {
```
Comment (matschreiner) [inline neural_lam/models/ar_model.py]: This code is hard to read. Maybe wrap it in a function with a descriptive name? I would imagine that the train_log_dict should be defined in one line, something like

`train_log_dict = {"train_loss": batch_loss, "lr": ..., **rmse_dict}`

File: neural_lam/models/ar_model.py [hunk 7]
```diff
@@ -298,9 +309,18 @@ def training_step(self, batch):
...
+            f"train_rmse_{v}": mean_rmse_ar_step_1[i]
+            for (i, v) in enumerate(state_var_names)
+        }
+        train_log_dict["train_lr"] = self.trainer.optimizers[0].param_groups[0][
```
Comment (matschreiner) [inline neural_lam/models/ar_model.py]: isn't learning rate only related to training since it's called "train_lr"?

File: neural_lam/models/ar_model.py [hunk 8]
```diff
@@ -343,6 +366,15 @@ def validation_step(self, batch, batch_idx):
             if step <= len(time_step_loss)
         }
         val_log_dict["val_mean_loss"] = mean_loss
```
Comment (matschreiner) [inline neural_lam/models/ar_model.py]: same as above

File: neural_lam/models/ar_model.py [hunk 9]
```diff
@@ -352,13 +384,6 @@ def validation_step(self, batch, batch_idx):
         )
 
         # Store MSEs
```
Comment (matschreiner) [inline neural_lam/models/ar_model.py]: Redundant comment

**Comment by matschreiner** [comment]
Sorry @mafdmi I added a lot of comments.

Nice that we are logging all the metrics now. I had hoped that we could find a common metric to compare models across runs with different variables.

**Comment by mfroelund** [comment]
> Sorry @mafdmi I added a lot of comments.
> 
> Nice that we are logging all the metrics now. I had hoped that we could find a common metric to compare models across runs with different variables.

I don't get it, but I still don't see your comments!

**Comment by mfroelund** [comment]
> @matschreiner I've fixed the tests, but for some reason two of the workflow jobs haven't been run - they have been waiting for a runner to pick them up since yesterday. Tried to re-run them today: https://github.com/mafdmi/neural-lam/actions/runs/13700752255

@leifdenby Do you know why the gpu tests fail in above action? Is it okay to merge without those tests passing?

File: neural_lam/models/ar_model.py (hunk 4 above)
Comment (mfroelund) [reply neural_lam/models/ar_model.py]: Agree, I've added a line on the mse calculation, but I don't think I know enough to reword the existing docstring.

File: neural_lam/models/ar_model.py (hunk 5 above)
Comment (mfroelund) [reply neural_lam/models/ar_model.py]: Agree, maybe somebody else can contribute here, since I don't know enough to properly phrase it.

File: neural_lam/models/ar_model.py (hunk 6 above)
Comment (mfroelund) [reply neural_lam/models/ar_model.py]: Not sure if it is more readable now, but I made it into one dict. I haven't put it into its own function, since I thought it is part of the training_step or validation_step to log. But we can do that if you think it'd be better.

File: neural_lam/models/ar_model.py (hunk 5 above)
Comment (matschreiner) [reply neural_lam/models/ar_model.py]: The problem is that this function is actually unpacking the batch and performing the unroll prediction step. Then it returns some elements of the unpacked batch and the calculation, then calculate metrics, which is an unclear responsibility that is hard to name and write a good docstring for and that is what happened. :D A cleaner implementation I would say is to just unpack the batch in the steps and perform the prediction. 

So basically remove the `common_step` function entirely and replace it by

```
(init_states, target_states, forcing_features, batch_times) = batch
prediction, pred_std = self.unroll_prediction(init_states, forcing_features, target_states)  
entry_mses = .... 
```

**Comment by matschreiner** [comment]
@mafdmi I had a last comment, otherwise LGTM

**Comment by mfroelund** [comment]
How does it look now to you? I've removed the common_steps function now. Makes sense I think:)

File: neural_lam/models/ar_model.py (hunk 7 above)
Comment (mfroelund) [reply neural_lam/models/ar_model.py]: I'm not sure what you mean.. actually I'm unsure if we have learning rate for test and validation steps? If not, yes, then let's simply call it "lr"
